### PR TITLE
emacs29: 29.1-rc1 -> 29.1

### PR DESCRIPTION
**Pull request by idlip**
## Description of changes

Changelog: https://git.savannah.gnu.org/cgit/emacs.git/tree/etc/NEWS?h=emacs-29

Release Notes: https://lists.gnu.org/archive/html/emacs-devel/2023-07/msg00879.html

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by idlip** [comment]
Note: I have not built the package, I have low spec hardware.

I just wanted to bump the update quickly, so I just changed the name and got the hash from running:

`nix-prefetch-url "https://git.savannah.gnu.org/cgit/emacs.git/snapshot/emacs-29.1.tar.gz"`

Let me if there is some mistake or some major changes, as to take more steps, rather than simple name change, so I can close this PR.

**Comment by wahjava** [comment]
> Note: I have not built the package, I have low spec hardware.
> 
> I just wanted to bump the update quickly, so I just changed the name and got the hash from running:
> 
> `nix-prefetch-url "https://git.savannah.gnu.org/cgit/emacs.git/snapshot/emacs-29.1.tar.gz"`
> 
> Let me know if there is some mistake or some major changes, as to take more steps, rather than simple name change, so I can close this PR.

For `hash` attribute, value needs to be in [SRI format](https://en.wikipedia.org/wiki/Subresource_Integrity), which you can [fetch using this](https://discourse.nixos.org/t/why-does-nix-prefetch-url-not-return-hashes-in-sri-format/18271/4).

**Comment by panchoh** [comment]
Also, don't forget to add the label `backport release-23.05`.

https://github.com/NixOS/nixpkgs/pull/245068#issuecomment-1647246675

File: pkgs/applications/editors/emacs/sources.nix [hunk 1]
```diff
@@ -50,12 +50,12 @@ in
...
     pname = "emacs";
-    version = "29.1-rc1";
+    version = "29.1";
     variant = "mainline";
     src = fetchFromSavannah {
       repo = "emacs";
-      rev = "29.1-rc1";
-      hash = "sha256-p0lBSKsHrFwYTqO5UVIF/PgiqwdhYQE4oUVcPtd+gsU=";
+      rev = "29.1";
+      hash = "1k6jkyyb9g0dsd4nf93hdf9d6nq7a0sqz26q6ls42c04zi7v9319";
```
Comment (panchoh) [inline pkgs/applications/editors/emacs/sources.nix:58]: `nix hash to-sri --type sha256 1k6jkyyb9g0dsd4nf93hdf9d6nq7a0sqz26q6ls42c04zi7v9319`
`sha256-KYy0T/wEMEE0NdiIjzVQB1vTkmtwJGdJ0w28tLyf0sw=`

Comment (panchoh) [reply pkgs/applications/editors/emacs/sources.nix:58]: Or:
`nix store prefetch-file "https://git.savannah.gnu.org/cgit/emacs.git/snapshot/emacs-29.1.tar.gz"`
Downloaded 'https://git.savannah.gnu.org/cgit/emacs.git/snapshot/emacs-29.1.tar.gz' to '/nix/store/3j07pifs2kh5yap9vky4vk3z4hfhcxc1-emacs-29.1.tar.gz' (hash '`sha256-KYy0T/wEMEE0NdiIjzVQB1vTkmtwJGdJ0w28tLyf0sw=`').

Comment (idlip) [reply pkgs/applications/editors/emacs/sources.nix]: Thank you for the info, definitely felt like I was missing that.

Changed the hash!

**Review by panchoh — APPROVED** [review]
LGTM

**Comment by zendo** [comment]
Also `all-packages.nix` should be change to `emacs = emacs29;`

**Comment by idlip** [comment]
Thank you for the clarification!

**Comment by panchoh** [comment]
> Also `all-packages.nix` should be change to `emacs = emacs29;`

But only on unstable, right?

**Comment by idlip** [comment]
> > Also all-packages.nix should be change to emacs = emacs29;

> But only on unstable, right?

Should I change that?

Then there is no wait for pkg.emacs29 (which currently exist on unstable?)
Also in that case, new addition as, is also expected:

`emacs-pgtk = emacs29-pgtk`

**Comment by panchoh** [comment]
> > > Also all-packages.nix should be change to emacs = emacs29;
> 
> > But only on unstable, right?
> 
> Should I change that?
> 
> Then there is no point for pkg.emacs29 (which currently exist on unstable?) Also in that case, new addition as, is also expected:
> 
> `emacs-pgtk = emacs29-pgtk`

The problem is that the version bump should apply to both `unstable` and `release-23.05`.
But if you add these other changes as well, then that's going to break `release-23.05`, since the release should not alter the (major) versions of the packages included.

I would do these changes in a separate PR, pertaining only to `unstable`.

But I'm a NixOS novice, so let's hear also to more experienced people.

Thanks for the PR, BTW!

**Comment by adisbladis** [comment]
Please don't change the main attribute in this PR. Let's just get the bump merged first and then do a follow-up changing the main attribute.
That requires release notes and more thorough testing.

**Comment by idlip** [comment]
I'm newbie at this matter lol, I just opened pr to notify and just change the hash.

Glad to contribute to nixos, and it is such a wonderful place.

Somebody, or you guys can continue with other steps/PR.

**Comment by adisbladis** [comment]
Emacs overlay](https://github.com/nix-community/emacs-overlay) auto-updated a couple of hours ago and we've been running hydra jobsets for 29.1 on the nix-community hydra in https://hydra.nix-community.org/eval/400298?filter=-unstable&compare=399847&full= with no visible regressions.

**Comment by AndersonTorres** [comment]
For now let's treat Emacs 28 as the default, changing it to 29 in the next release.

**Comment by adisbladis** [comment]
```
         specified: sha256-KYy0T/wEMEE0NdiIjzVQB1vTkmtwJGdJ0w28tLyf0sw=
            got:    sha256-3HDCwtOKvkXwSULf3W7YgTz4GV8zvYnh2RrL28qzGKg=
```

@idlip While I appreciate your enthusiasm you need to apply a bit of testing before opening a PR.
This is too low of an effort to meet our bar. Next time at the very least build the derivation that you want to contribute.

While not all checkboxes in the PR template are mandatory you should at least check _some_ of them.

I am going to close this one and open my own bump PR as you already state that you are unable to build anything.

**Comment by idlip** [comment]
Sure, no worries. I understand, and thank you for taking your time on this!

Sorry for the pings and getting your attention.

Next time you see, hopefully I will be upgraded to next version 😅.

**Comment by adisbladis** [comment]
https://github.com/NixOS/nixpkgs/pull/246203

**Comment by panchoh** [comment]
> ```
>          specified: sha256-KYy0T/wEMEE0NdiIjzVQB1vTkmtwJGdJ0w28tLyf0sw=
>             got:    sha256-3HDCwtOKvkXwSULf3W7YgTz4GV8zvYnh2RrL28qzGKg=
> ```
> 
> @idlip While I appreciate your enthusiasm you need to apply a bit of testing before opening a PR. This is too low of an effort to meet our bar. Next time at the very least build the derivation that you want to contribute.
> 
> While not all checkboxes in the PR template are mandatory you should at least check _some_ of them.
> 
> I am going to close this one and open my own bump PR as you already state that you are unable to build anything.

TIL: The output of `nix store prefetch-file "url"` (or `nix-prefetch-url "url"`, for that matter), don't produce the same hash (assuming same encoding of the hash) as the evaluation of the derivation. It seems that `nix` does not just hash the downloaded `.tar.gz`, it first expands the archive, normalizes it, and then obtains the hash of the tree.

To sum up, there is no alternative to a) bump the version and b) run `nixos-rebuild test or switch` and obtain the expected hash from there.

Again, I'll leave up to more experienced people to comment on this.

**Comment by peterhoeg** [comment]
> To sum up, there is no alternative to a) bump the version and b) run nixos-rebuild test or switch and obtain the expected hash from there.

Well, you can just build the package without having to eval the entire nixos. 

```sh
cd $WHEREEVER_YOU_CLONED_NIXPKGS
git checkout -b my_local_version upstream/nixos-unstable # less rebuilding when working against unstable compared to master
# now bump the version, and then
nix build -f . emacs-29  # build whatever thing you are working on - it will show you the proper hash here
```

Much faster than running `nixos-rebuild`.

**Comment by idlip** [comment]
I did not know that, Thank you for the tip.

TIL.

**Comment by AndersonTorres** [comment]
I will write a tutorial about this.
Nonetheless, everything was already explained on the wiki:

https://nixos.wiki/wiki/Nixpkgs/Create_and_debug_packages

**Pull request closed without merging**